pkg/serviceaccounttoken/secret_cleaner.go

```diff
@@ -0,0 +1,230 @@
+package serviceaccounttoken
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	clientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	clientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const (
+	// this can't be imported from the impersonation namespace because of a
+	// cycle.
+	impersonationNamespace string = "cattle-impersonation-system"
+
+	pageSize int64 = 100
+)
+
+var (
+	// How long do we wait before the first secret cleanup run.
+	cleanCycleDelay time.Duration = time.Second * 5
+	// How often do we clean secrets?
+	cleanCycleTime time.Duration = time.Minute * 1
+
+	// How many secrets are fetched at a time?
+	// Not all secrets are removed from this list.
+	cleaningBatchSize int64 = 1000
+)
+
+// StartServiceAccountSecretCleaner starts a background process to cleanup old
+// service accounts secrets.
+//
+// This should only be started in the leader pod.
+func StartServiceAccountSecretCleaner(ctx context.Context, client clientcorev1.CoreV1Interface) error {
+	if strings.ToLower(os.Getenv("DISABLE_SECRET_CLEANER")) == "true" {
+		logrus.Info("ServiceAccountSecretCleaner disabled - not starting")
+		return nil
+	}
+
+	secrets := client.Secrets(impersonationNamespace)
```

Comment:
1. it is my understanding this code is executed both in upstream and all downstream clusters - `client` will be a client on the particular cluster the code is run in. Is this correct, and expected?
2. IIUC in the case of the upstream cluster, we have a complete Secret cache available (as opposed to a client), and for downstream caches we do not have a complete cache but we do have a complete cache of the impersonation namespace at least (see https://github.com/rancher/rancher/pull/46982)

If 2. is correct, shouldn't we use the existing cache instead of a client? That would, among other things, avoid the problem of pagination as we already have the whole data set in memory.

_(yes, one day we might want to remove that cache, but it's not decided yet)_